### PR TITLE
Partition all-cabal-hashes into case-insensitive-safe components.

### PR DESCRIPTION
**Pull request by shlevy**
OS X by default has a case-insensitive filesystem, and fetching
all-cabal-hashes there fails due to a hash mismatch caused by package
pairs like compactable and Compactable. This partitions the package set
such that each partition contains no equivalent-up-to-case pairs.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by shlevy** [comment]
Have an idea for simplifying this and hopefully speeding it up significantly.

**Comment by shlevy** [comment]
OK, recent version is much faster. Faster than a `tar xf`, even.

**Comment by shlevy** [comment]
Barring any objection I'll merge mid next week @peti

File: pkgs/data/misc/hackage/default.nix [hunk 1]
```diff
@@ -1,11 +1,33 @@
...
-  owner = "commercialhaskell";
-  repo = "all-cabal-hashes";
-  rev = "901c2522e6797270f5ded4495b1a529e6c16ef45";
-  sha256 = "0wng314y3yn6bbwa5ar254l7p8y99gsvm8ll4z7f3wg77v5fzish";
-}
+let partition-all-cabal-hashes = stdenv.mkDerivation
+      { name = "partition-all-cabal-hashes";
+        src = ./partition-all-cabal-hashes.c;
+        unpackPhase = "true";
+        buildInputs = [ pkgconfig libarchive glib ];
```
[the remark this answers is not in the document]
Comment (shlevy) [reply pkgs/data/misc/hackage/default.nix]: @Ericson2314 I'll make that change, but FWIW this is almost always going to be used at eval time and thus not cross-compiled.

**Review by peti — COMMENTED** [review]
The functionality provided by partition-all-cabal-hashes.c seems more general, i.e. I assume that many other projects have similar problems. Would it be possible to factor that tool out of `pkgs/data/misc/hackage` somehow?

Generally speaking, I'll drop all use of the "all-cabal-hashes" repo ASAP since this stuff if broken in several other ways, too, and we can find the hashes in the 01-index.tar.gz tarball from cabal-install anyway these days.

**Comment by shlevy** [comment]
Hmm I don't know of any other projects with similar problems. Anyway this assumes things about the structure of the tarball, it could possibly be generalized but I would like to see another example or two to see how it can be done usefully.

What is the structure of the 01-index.tar.gz tarball?

**Comment by shlevy** [comment]
Just checked that tarball, it will have exactly the same issue and this same tool will work on it.

**Comment by peti** [comment]
> Just checked that tarball, it will have exactly the same issue.

No, it won't, because we'll never extract it.

**Comment by shlevy** [comment]
Ah, cabal2nix will extract info directly from the tarball? Nice.

Can we merge this in the mean time though?

**Comment by peti** [comment]
If it works, then I don't see why not.

**Comment by ElvishJerricco** [comment]
@shlevy Wouldn’t it have been easier to just extract individual files from the tar file as needed?

**Comment by shlevy** [comment]
@ElvishJerricco Yeah, fair point :grinning:

**Pull request merged**